### PR TITLE
6080 improve changelog verification

### LINKED ISSUE #6080: Validate changelog changes in PRs
Refs: https://github.com/mapbox/mapbox-navigation-android/pull/6072#issuecomment-1189493804

Now we have the changelog verification build that checks just the fact that CHANGELOG was updated. But with the current approach we can easily put the line in the wrong place if the PR was open, then there was a release, then the PR was merged.
Suggested solution is to check not only that the file was updated, but also the following:
1. Added lines are in the Unreleased section;
2. Entries are not duplicated across multiple subsequent stable releases with incremental version (they can still be duplicated between pre-releases and stable releases).

Changelog verification happens [here](https://github.com/mapbox/mapbox-navigation-android/blob/main/scripts/validate-changelog.py).

### PR DESCRIPTION
**Pull request by dzinad**
Closes #6080.

1. Fixed the PR link check: before it checked that there's a PR link in the diff, while diff can be sth like:
```
diff --git a/CHANGELOG.md b/CHANGELOG.md
index d5960be9d68..4a5bd3d04af 100644
--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,6 +4,7 @@ Mapbox welcomes participation and contributions from everyone.
 
 ## Unreleased
 #### Features
+- Added ddtest.
 - Added `ViewBinderCustomization.infoPanelBinder` that allows installation of a custom info panel layout in `NavigationView`. [#6049](https://github.com/******/******-navigation-android/pull/6049) 
 - Added `ViewStyleCustomization.infoPanelPeekHeight` that allows customization of `NavigationView` info panel bottom sheet peek height. [#6049](https://github.com/******/******-navigation-android/pull/6049) 
 - Added `ViewStyleCustomization.infoPanelMarginStart`, `ViewStyleCustomization.infoPanelMarginEnd`, `ViewStyleCustomization.infoPanelBackground` that allows customization of a `NavigatioView` default info panel margins and background. [#6049](https://github.com/******/******-navigation-android/pull/6049)
```
So it didn't actually check the added lines.
2. Added the check to verify that the added line is in the 'Unreleased' section.
3. Added the check to verify that the added line is not contained changelog for any other _stable_ version.

Notes:
For now I left one "skip-changelog" label for everything. It disables all checks. If anyone thinks we need more precise labels (to skip the first check, or the second one, etc.), let's discuss.

### CODE REVIEW

**Comment by VysotskiVadim** [comment]
@dzinad , what do you think about adding some tests which verifies logic of analysing changelog diff? 

I do not suggest running those tests on CI, I'm thinking about one more `py` file which tests complex functions from `scripts/validate-changelog.py` so that I can run tests locally when I edit the script.

**Comment by dzinad** [comment]
> @dzinad , what do you think about adding some tests which verifies logic of analysing changelog diff?
I do not suggest running those tests on CI, I'm thinking about one more py file which tests complex functions from scripts/validate-changelog.py so that I can run tests locally when I edit the script.

Added some.

File: scripts/validate_changelog_utils.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    unreleased_group = extract_unreleased_group(versions)
+    stable_versions = extract_stable_versions(versions)
+
+    for added_line in added_lines:
+        if added_line not in unreleased_group:
+            raise Exception(added_line + " should be placed in 'Unreleased' section")
+
+        for stable_version in stable_versions:
+            if added_line in stable_versions[stable_version]:
+                raise Exception("The changelog entry \"" + added_line + "\" is already contained in " + stable_version + " changelog.")
```
Comment (VysotskiVadim) [inline scripts/validate_changelog_utils.py:108]: Does it mean that I won't be able to change changelog of released version?
If yes, maybe it's worth adding a "modify-old-changelog" label to allow modification of old changelog? It can be useful if you want to fix a typo or clarify something after a question from a customer? What do you think?

File: scripts/validate_changelog_utils.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            stable_versions[version_name] = versions[version]
+    return stable_versions
+
+def should_skip_changelog(response_json):
+    if "labels" in response_json:
+        pr_labels = response_json["labels"]
+        for label in pr_labels:
+            if label["name"] == "skip changelog":
+                return True
+    return False
```
Comment (VysotskiVadim) [inline scripts/validate_changelog_utils.py:62]: maybe we shouldn't skip changelog checks if changelog file is modified?

**Review by VysotskiVadim — APPROVED** [review]
Nice improvement 👍 

Dropped a few comments about labels. What do you think about more complex system for labels:
* "Skip changelog" - verifies that changelog file wasn't modified
* "Modified changelog" - lets you modify existing release, but script verifies links
* "Ignore changelog erros" - turns off all checks

?

BTW, this could be an overkill, current solution covers main flow, consider it as an idea 🙂

File: scripts/validate_changelog_utils.py (hunk 1 above)
Comment (dzinad) [reply scripts/validate_changelog_utils.py:108]: You're right. Makes sense, however, I don't want to overcomplicate it. If you specifically make a PR that clarifies an old entry, you just disable the check. This check is mostly introduced for two reasons: 1) so that we don't forget to update CHANGELOG.md; 2) so that our lines do not appear in random places after rebase after release.
I want to keep the system as simple as possible. For special cases it's absolutely OK to check turn the check off.

**Comment by dzinad** [comment]
> What do you think about more complex system for labels:
"Skip changelog" - verifies that changelog file wasn't modified
"Modified changelog" - lets you modify existing release, but script verifies links
"Ignore changelog erros" - turns off all checks
?
BTW, this could be an overkill, current solution covers main flow, consider it as an idea 🙂

Yes, I think it's an overkill. More details [here](https://github.com/mapbox/mapbox-navigation-android/pull/6083#discussion_r930120575).

File: scripts/validate_changelog_utils.py (hunk 2 above)
Comment (dzinad) [reply scripts/validate_changelog_utils.py:62]: I think historically this label means "skip changelog checks". Anyways, I think it should mean this now. Maybe we should just rename it?
And we should skip the checks if it was modified and the label is present. Because there are more checks now and we can ignore, as you described, fixing a typo or clarifying an entry. For the simplicity of it I'd like to keep one label only.

File: scripts/test_resources/test_changelog.md [hunk 3]
```diff
@@ -0,0 +1,3757 @@
+# Changelog for the Mapbox Navigation SDK for Android
```
Comment (LukasPaczos) [inline scripts/test_resources/test_changelog.md:1]: Could we prepare a more abstract (and shorter) test cases? I'd like to avoid duplicates in search queries and just save on pure size to maintain.
